File: algorithm/java/backjoon/bracket.java
package backjoon;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

class bracket {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int num = Integer.parseInt(br.readLine());

        for (int i=0; i < num; i++) {
            String bra = br.readLine();
            Stack<Character> stack = new Stack<>();

            for (int j=0; j < bra.length(); j++) {
                char c = bra.charAt(j);

                if (c == '(') {
                    stack.push(c);
                } else if (stack.empty()) {
                    stack.push(c);
                    break;
                } else {
                    stack.pop();
                }
            }

            if (stack.empty()) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
    }
}

File: algorithm/python/backjoon/stack/bracket.py
import sys
In = sys.stdin.readline
N = int(In())

for i in range(N):
    bracket = In().strip() # int로 바꾸거나 .strip()을 하거나
    stack = []
    for j in range(len(bracket)):
        if bracket[j] == "(":
            stack.append(bracket[j])
            print(stack)
        elif len(stack) == 0:
            stack.append(bracket[j])
            print(stack)
            break
        else:
            stack.pop()
            print(stack)

    # if stack:
    #     print("NO")
    # else:
    #     print("YES")